sui_programmability/examples/nfts/sources/DiscountCoupon.move

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+module NFTs::DiscountCoupon {
+    use Sui::Coin;
+    use Sui::NFT::{Self, NFT};
+    use Sui::SUI::SUI;
+    use Sui::Transfer;
+    use Sui::TxContext::{Self, TxContext};
+
+    /// Sending to wrong recipient.
+    const EWRONG_RECIPIENT: u64 = 0;
+
+    /// Percentage discount out of range.
+    const EOUT_OF_RANGE_DISCOUNT: u64 = 1;
+
+    /// Discount coupon NFT.
+    struct DiscountCoupon has store {
+        // coupon issuer
+        issuer: address,
+        // percentage discount [1-100]
+        discount: u8,
+        // expiration timestamp (UNIX time) - app specific
+        expiration: u64,
+    }
+
+    /// Simple issuer getter.
+    public fun issuer(coupon: &DiscountCoupon): address {
+        coupon.issuer
+    }
+
+    /// Mint then transfer a new `DiscountCoupon` NFT, and top up recipient with some SUI.
+    public fun mint_and_topup(
+        discount: u8,
+        expiration: u64,
+        recipient: address,
+        coin: Coin::Coin<SUI>, 
+        ctx: &mut TxContext,
+    ) {
+        assert!(discount > 0 && discount <= 100, EOUT_OF_RANGE_DISCOUNT);
+        let nft = NFT::mint(
+                DiscountCoupon { 
+                    issuer: TxContext::sender(ctx),
+                    discount: discount, 
+                    expiration: expiration,
+                }, 
+                ctx);
+        Transfer::transfer(nft, recipient);
+        Sui::SUI::transfer(coin, recipient, ctx);
+    }
+
+    /// Burn DiscountCoupon.
+    public fun burn(nft: NFT<DiscountCoupon>, _ctx: &mut TxContext) {
+        let DiscountCoupon { issuer: _, discount: _, expiration: _ } = NFT::burn(nft);
+    }
+
+    /// Transfer DiscountCoupon to issuer only.
+    //  TODO: Consider adding more valid recipients. 
+    //      If we stick with issuer-as-receiver only, then `recipient` input won't be required).
+    public fun transfer(nft: NFT<DiscountCoupon>, recipient: address, _ctx: &mut TxContext) {
```

Comment:
Might be better called `return`, given it's not generic transfer?
Why do we need this API btw? i.e. what incentive does a person have to return this coupon?

Reply:
The PoC requirement is coupons shouldn't be transferred between users, but only return them back to the issuer (or in the future to some merchant allowlist), an action which app-wise will be considered as using / cashing the NFT to the shop.